[API] Refactor: core cache design and impl

Xunzhuo opened the pull request:
## Pull Request Description

refactor: core cache desgin and impl

---

<details>
<summary><strong>Contribution Guidelines (Expand for Details)</strong></summary>

<p>We appreciate your contribution to aibrix! To ensure a smooth review process and maintain high code quality, please adhere to the following guidelines:</p>

<h3>Pull Request Title Format</h3>
<p>Your PR title should start with one of these prefixes to indicate the nature of the change:</p>
<ul>
    <li><code>[Bug]</code>: Corrections to existing functionality</li>
    <li><code>[CI]</code>: Changes to build process or CI pipeline</li>
    <li><code>[Docs]</code>: Updates or additions to documentation</li>
    <li><code>[API]</code>: Modifications to aibrix's API or interface</li>
    <li><code>[CLI]</code>: Changes or additions to the Command Line Interface</li>
    <li><code>[Misc]</code>: For changes not covered above (use sparingly)</li>
</ul>
<p><em>Note: For changes spanning multiple categories, use multiple prefixes in order of importance.</em></p>

<h3>Submission Checklist</h3>
<ul>
    <li>[ ] PR title includes appropriate prefix(es)</li>
    <li>[ ] Changes are clearly explained in the PR description</li>
    <li>[ ] New and existing tests pass successfully</li>
    <li>[ ] Code adheres to project style and best practices</li>
    <li>[ ] Documentation updated to reflect changes (if applicable)</li>
    <li>[ ] Thorough testing completed, no regressions introduced</li>
</ul>

<p>By submitting this PR, you confirm that you've read these guidelines and your changes align with the project's contribution standards.</p>

</details>

varungup90 commented:
@zhangjyr is also working on same implementation.

Xunzhuo commented:
@varungup90 is there a PR for it?

Jeffwan commented:
@Xunzhuo I think jingyuan is working on hetegenous routing and working on extracting the refactor related logics. Here's his RFC issue https://github.com/vllm-project/aibrix/issues/868

Xunzhuo commented:
This refactor does not break the previous usage for cache pkg, just improve the code maintainability, which I think is not conflict with the #868.

Jeffwan commented:
@Xunzhuo that's even better. I will take a look tomorrow

Xunzhuo commented:
CI has passed, plz help review @varungup90 @Jeffwan

Jeffwan commented:
@Xunzhuo I reviewed the PR, and the refactor looks straightforward with a few key changes:

1. Extracting a root-level interface: Introduced a`Cache` interface with `PodCache`, `ModelCache`, `MetricCache` and `TraceCache` interfaces.
2. Refactoring `cache.go`: Split the lengthy cache.go file into multiple, more manageable files
3. Method renaming: Updated method names for clarity, e.g., `GetModelsForPod` -> `ListModelsByPod` etc.

These changes improve maintainability, especially for new contributors. Let me know if there are any specific areas you'd like me to focus on. BTW, Feel free to write key changes in the PR description next time.

zhangjyr commented:
Thanks for the PR. I have reviewed the new interfaces. Decoupling different cache data providers is necessary. However, PodCache and ModelCache should be unified as one interface, as they solely depend on pod information. We should maintain the atomicity of updating pod metadata, model(adapter) metadata, and their relation.

Another thing I need to clarify is that RequestCount series methods store real-time workload statistics, which are distinguished from cached metrics pulled from other data sources. I would suggest either:
1. Using cached metrics as a store and implementing real-time RequestCount updates in extended metrics providers.
2. Wrap RequestCount logic to a new real-time metrics provider.

The cache and routing part will be refactored to improve concurrent access in #884

Jeffwan commented:
> However, PodCache and ModelCache should be unified as one interface, as they solely depend on pod information. We should maintain the atomicity of updating pod metadata, model(adapter) metadata, and their relation.

@zhangjyr The informer code remains the same, I think it's still pod change driven. PodCache and ModelCache are two logical interfaces that only provides Get/List capabilities.  Atomicity should be good.

Xunzhuo commented:
> @Xunzhuo I reviewed the PR, and the refactor looks straightforward with a few key changes:
> 
> 
> 
> 1. Extracting a root-level interface: Introduced a`Cache` interface with `PodCache`, `ModelCache`, `MetricCache` and `TraceCache` interfaces.
> 
> 2. Refactoring `cache.go`: Split the lengthy cache.go file into multiple, more manageable files
> 
> 3. Method renaming: Updated method names for clarity, e.g., `GetModelsForPod` -> `ListModelsByPod` etc.
> 
> 
> 
> These changes improve maintainability, especially for new contributors. Let me know if there are any specific areas you'd like me to focus on. BTW, Feel free to write key changes in the PR description next time. 
> 


Thanks, it's most of the refactor key points. 

I will create a clear issue and PR description in later works.

zhangjyr commented:
> > However, PodCache and ModelCache should be unified as one interface, as they solely depend on pod information. We should maintain the atomicity of updating pod metadata, model(adapter) metadata, and their relation.
> 
> @zhangjyr The informer code remains the same, I think it's still pod change driven. PodCache and ModelCache are two logical interfaces that only provides Get/List capabilities. Atomicity should be good.

@Jeffwan I see your point. The proposal interfaces only defined the read operations. The data provisioning currently is still intact. I think we can merge the PR after #884 merged.

Jeffwan commented:
@zhangjyr I feel this one is straightforward, PR #884 has more feature refactors, let's merge this one first? WDYT?
